### PR TITLE
GPU memory-intensive codegen: add column reduction schedule.

### PR DESCRIPTION
**Pull request by yunzhongOvO**
This PR is the optimized column reduction schedule in GPU memory-intensive code generation flow. 
Schedule 1: thread_num threads load thread_num * tile_h elements each thread reduces tile_h elements from gmem and atomic to gmem.
Schedule 2: block_x * block_y threads load block_x * block_y * tile_h elements, each thread reduces tile_h elements from gmem to SHM(block_x * block_y),  do reduction in SHM and atomic to gmem.

### CODE REVIEW

**Comment by qiuxiafei** [comment]
Please add some UT~

**Comment by yunzhongOvO** [comment]
> 

Done.

**Comment by yunzhongOvO** [comment]
> 

Done. UTs are added.

File: tao_compiler/mlir/disc/transforms/lhlo_legalize_roots_to_loops.cc [hunk 1]
```diff
@@ -503,342 +503,546 @@ void emitNotToVectorReduction(OpBuilder& b, Location loc, Operation* root_op,
...
+//         }
+//       }
+//     } else {
+//       accum = init_value;
+//     }
+//     shm[shm_trans_index] = accum;
+//     for (int stride = num_threads_row / 2; stride > 0; stride /= 2) {
+//       __syncthreads();
+//       if (thread_row_index < stride) {
+//         sdata[shm_trans_index] += sdata[shm_trans_index + stride];
```
Comment (JamesTheZ) [inline tao_compiler/mlir/disc/transforms/lhlo_legalize_roots_to_loops.cc:771]: Use the same name for `shm` and `sdata`.

**Comment by yunzhongOvO** [comment]
> Please add some UT~

Done.

File: tao_compiler/mlir/disc/transforms/lhlo_legalize_roots_to_loops.cc [hunk 2]
```diff
@@ -503,342 +503,546 @@ void emitNotToVectorReduction(OpBuilder& b, Location loc, Operation* root_op,
...
+  std::copy_if(
+      root_ops.begin(), root_ops.end(), std::back_inserter(col_reduction_roots),
+      [](Operation* operation) { return isRank2ColReduction(operation); });
+
+  Value lhs = *dominant_op->getOperands().begin();
+  const int kThreads_col = THREADS_COL;              // 32
+  const int kThreads_row = THREADS_ROW;              // 8
+  const int kTileH = TILE_H;                         // 64
+  const int kTileRow = kThreads_row * TILE_H;        // 512
+  const int kThreads = kThreads_col * kThreads_row;  // 256
```
Comment (JamesTheZ) [inline tao_compiler/mlir/disc/transforms/lhlo_legalize_roots_to_loops.cc:802]: The comments about the value could be removed if they are not fixed.

File: tao_compiler/mlir/disc/transforms/lhlo_legalize_roots_to_loops.cc [hunk 3]
```diff
@@ -503,342 +503,546 @@ void emitNotToVectorReduction(OpBuilder& b, Location loc, Operation* root_op,
...
+  Value shm_offset1 = b.create<arith::AddIOp>(loc, shm_trans_index, one);
+  for (auto root_pair : llvm::enumerate(col_reduction_roots)) {
+    Operation* root_op = root_pair.value();
+    int idx = root_pair.index();
+    auto root_element_type = getLhloOpsElementType(root_op);
+    Value shm_op0 =
+        b.create<memref::LoadOp>(loc, shared_mem_map[root_op], shm_trans_index);
+    Value shm_op1 =
+        b.create<memref::LoadOp>(loc, shared_mem_map[root_op], shm_offset1);
+    Value partial_result = (accum_factory[idx])(shm_op0, shm_op1);
```
Comment (JamesTheZ) [inline tao_compiler/mlir/disc/transforms/lhlo_legalize_roots_to_loops.cc:1035]: Is this more effective to calculate the partial_result here than in `for (int stride = num_threads_row / 2; stride > 0; stride /= 2)`?

File: tao_compiler/mlir/disc/transforms/codegen_utils.h [hunk 4]
```diff
@@ -97,6 +99,7 @@ constexpr const int kThreadsRowReduction = 512;
 #else
 constexpr const int kThreadsRowReduction = 256;
 #endif
+constexpr const int kThreadsRowReduction512 = 512;
```
[the remark this answers is not in the document]
Comment (yunzhongOvO) [reply tao_compiler/mlir/disc/transforms/codegen_utils.h:102]: Done.

File: tao_compiler/mlir/disc/transforms/codegen_utils.h [hunk 5]
```diff
@@ -83,6 +83,8 @@ using DiscColReductionScheduleType = enum : int {
   DISC_TILE_LOOP_W64_H8 = 4,
   DISC_TILE_LOOP_W16_H32 = 5,
   DISC_TILE_LOOP_W8_H8 = 6,
+  DISC_FLAT = 7,
+  DISC_THIN = 8,
```
[the remark this answers is not in the document]
Comment (yunzhongOvO) [reply tao_compiler/mlir/disc/transforms/codegen_utils.h]: Done.

File: tao_compiler/mlir/disc/transforms/lhlo_legalize_roots_to_loops.cc [hunk 6]
```diff
@@ -503,342 +503,546 @@ void emitNotToVectorReduction(OpBuilder& b, Location loc, Operation* root_op,
...
+      root_ops.begin(), root_ops.end(), std::back_inserter(col_reduction_roots),
+      [](Operation* operation) { return isRank2ColReduction(operation); });
+
   Value lhs = *dominant_op->getOperands().begin();
-  const MemRefType& lhs_type = lhs.getType().template cast<MemRefType>();
-  const Type& element_type = lhs_type.getElementType();
-  const int c_tilesize_h = 128;
-  const int c_tilesize_w = 2;
+  const int kTileH = TILE_H;        // 32
+  const int kThreads = THREAD_NUM;  // 512
```
[the remark this answers is not in the document]
Comment (yunzhongOvO) [reply tao_compiler/mlir/disc/transforms/lhlo_legalize_roots_to_loops.cc]: Done.

File: tao_compiler/mlir/disc/transforms/lhlo_legalize_roots_to_loops.cc (hunk 1 above)
Comment (yunzhongOvO) [reply tao_compiler/mlir/disc/transforms/lhlo_legalize_roots_to_loops.cc]: Done.

File: tao_compiler/mlir/disc/transforms/lhlo_legalize_roots_to_loops.cc (hunk 2 above)
Comment (yunzhongOvO) [reply tao_compiler/mlir/disc/transforms/lhlo_legalize_roots_to_loops.cc]: Done.

File: tao_compiler/mlir/disc/transforms/lhlo_legalize_roots_to_loops.cc (hunk 3 above)
Comment (yunzhongOvO) [reply tao_compiler/mlir/disc/transforms/lhlo_legalize_roots_to_loops.cc]: Here I referred to the implementations of row reduction schedules and there is no performance gain. Now switched to `for (int stride = num_threads_row / 2; stride > 0; stride /= 2)`.

**Review by JamesTheZ — APPROVED** [review]
LGTM

**Pull request merged**